src/ucp/core/ucp_ep.c

```diff
@@ -3253,24 +3254,78 @@ ucp_ep_is_am_keepalive(ucp_ep_h ep, ucp_rsc_index_t rsc_idx)
     }
 
     wiface = ucp_worker_iface(ep->worker, rsc_idx);
-    return ucs_test_all_flags(wiface->attr.cap.flags,
-                              UCT_IFACE_FLAG_CONNECT_TO_IFACE |
-                              UCT_IFACE_FLAG_AM_BCOPY);
+    return (lane != UCP_NULL_LANE) &&
+           !(ucp_ep_config(ep)->p2p_lanes & UCS_BIT(lane)) &&
+           (wiface->attr.cap.flags & UCT_IFACE_FLAG_AM_BCOPY);
 }
 
-ucs_status_t ucp_ep_do_uct_ep_keepalive(ucp_ep_h ucp_ep, uct_ep_h uct_ep,
-                                        ucp_rsc_index_t rsc_idx, unsigned flags,
+static UCS_F_ALWAYS_INLINE uct_ep_h
+ucp_ep_select_keepalive(ucp_ep_h ucp_ep, ucp_rsc_index_t *rsc_idx_p,
+                        ucp_lane_index_t *lane_p)
+{
+    ucp_lane_index_t lane      = ucp_ep_config(ucp_ep)->key.keepalive_lane;
+    uct_ep_h uct_ep            = ucp_ep->uct_eps[lane];
```

Comment:
can init it under the `else` below, it is reinitialized for wireup

Reply:
also can do:
```
if (wireup_ep == NULL) {
    ...
    goto out;
}

if (wireup_ep->flags & UCP_WIREUP_EP_FLAG_REMOTE_CONNECTED) {
...
```

Reply:
I thought about it, but then we should initialize wireup_ep as `ucp_wireup_ep(ucp_ep->uct_eps[ucp_ep_config(ucp_ep)->key.keepalive_lane])` which is pretty long and better to save `lane`/`uct_ep` and then reinitialize them